File: pippip-native-ios/Support/Configurator.m
//
//  Configurator.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 2/12/18.
//  Copyright © 2018 seComm. All rights reserved.
//

#import "Configurator.h"
#import "ApplicationSingleton.h"
#import "AccountConfig.h"
#import "CKGCMCodec.h"
#import <Realm/Realm.h>

static NSLock *idLock = nil;

@interface Configurator ()
{
    NSMutableArray *privateWhitelist;
    NSMutableDictionary<NSString*, NSNumber*> *idMap;
}

@property (weak, nonatomic) SessionState *sessionState;

@end

@implementation Configurator

- (instancetype)init {
    self = [super init];

    privateWhitelist = [NSMutableArray array];
    _whitelist = privateWhitelist;
    idMap = [NSMutableDictionary dictionary];

    return self;

}

- (void)addContactId:(NSInteger)contactId withPublicId:(NSString *)publicId {

    NSInteger cid = [self getContactId:publicId];
    if (cid == NSNotFound) {
        idMap[publicId] = [NSNumber numberWithInteger:contactId];
        [self encodeIdMap:[self getConfig]];
    }

}

- (BOOL)addWhitelistEntry:(NSDictionary *)entity {

    AccountConfig *config = [self getConfig];
    if (privateWhitelist.count == 0) {
        [self decodeWhitelist:config];
    }
    NSString *publicId = entity[@"publicId"];
    NSUInteger idx = [privateWhitelist indexOfObjectPassingTest:^BOOL(id obj, NSUInteger index, BOOL *stop) {
        NSDictionary *entry = obj;
        return [publicId isEqualToString:entry[@"publicId"]];
    }];
    if (idx == NSNotFound) {
        [privateWhitelist addObject:entity];
        [self encodeWhitelist:config];
    }
    return idx == NSNotFound;

}

- (NSArray*)allContactIds {

    if (idMap.count == 0) {
        [self decodeIdMap:[self getConfig]];
    }
    return [idMap allValues];

}

- (void)decodeIdMap:(AccountConfig*)config {

    if (config.idMap != nil) {
        CKGCMCodec *codec = [[CKGCMCodec alloc] initWithData:config.idMap];
        NSError *error = nil;
        [codec decrypt:_sessionState.contactsKey withAuthData:_sessionState.authData withError:&error];
        if (error == nil) {
            NSInteger count = [codec getInt];
            while (idMap.count < count) {
                NSString *publicId = [codec getString];
                idMap[publicId] = [NSNumber numberWithInteger:[codec getInt]];
            }
        }
        else {
            NSLog(@"Error decoding contact ID map: %@", error.localizedDescription);
        }
    }

}

- (void)decodeWhitelist:(AccountConfig*)config {

    if (config.whitelist != nil) {
        CKGCMCodec *codec = [[CKGCMCodec alloc] initWithData:config.whitelist];
        NSError *error = nil;
        [codec decrypt:_sessionState.contactsKey withAuthData:_sessionState.authData withError:&error];
        if (error == nil) {
            NSInteger count = [codec getInt];
            while (privateWhitelist.count < count) {
                NSMutableDictionary *entity = [NSMutableDictionary dictionary];
                NSString *nickname = [codec getString];
                if (nickname.length > 0) {
                    entity[@"nickname"] = nickname;
                }
                entity[@"publicId"] = [codec getString];
                [privateWhitelist addObject:entity];
            }
        }
        else {
            NSLog(@"Error decoding whitelist: %@", error.localizedDescription);
        }
    }

}

- (void)deleteContactId:(NSString *)publicId {

    [idMap removeObjectForKey:publicId];
    [self encodeIdMap:[self getConfig]];

}

- (BOOL)deleteWhitelistEntry:(NSString *)publicId {

    AccountConfig *config = [self getConfig];
    if (privateWhitelist.count == 0) {
        [self decodeWhitelist:config];
    }
    NSUInteger deleteIndex = [privateWhitelist indexOfObjectPassingTest:^BOOL(id obj, NSUInteger index, BOOL *stop) {
        NSDictionary *entity = obj;
        return [publicId isEqualToString:entity[@"publicId"]];
    }];
    if (deleteIndex != NSNotFound) {
        [privateWhitelist removeObjectAtIndex:deleteIndex];
        [self encodeWhitelist:config];
    }
    return deleteIndex != NSNotFound;

}

- (void)encodeIdMap:(AccountConfig*)config {

    RLMRealm *realm = [RLMRealm defaultRealm];
    if (idMap.count > 0) {
        CKGCMCodec *codec = [[CKGCMCodec alloc] init];
        [codec putInt:idMap.count];
        for (NSString* publicId in idMap.allKeys) {
            [codec putString:publicId];
            NSNumber *cid = idMap[publicId];
            [codec putInt:[cid integerValue]];
        }
        NSError *error = nil;
        NSData *encoded = [codec encrypt:_sessionState.contactsKey
                            withAuthData:_sessionState.authData
                               withError:&error];
        if (error == nil) {
            [realm beginWriteTransaction];
            config.idMap = encoded;
            [realm commitWriteTransaction];
        }
        else {
            NSLog(@"Error while encoding contact ID map: %@", error.localizedDescription);
        }
    }
    else {
        [realm beginWriteTransaction];
        config.idMap = nil;
        [realm commitWriteTransaction];
    }
    
}

- (void)encodeWhitelist:(AccountConfig*)config {

    RLMRealm *realm = [RLMRealm defaultRealm];
    if (privateWhitelist.count > 0) {
        CKGCMCodec *codec = [[CKGCMCodec alloc] init];
        [codec putInt:privateWhitelist.count];
        for (NSDictionary *entity in privateWhitelist) {
            NSString *nickname = entity[@"nickname"];
            if (nickname == nil) {
                nickname = @"";
            }
            [codec putString:nickname];
            [codec putString:entity[@"publicId"]];
        }
        NSError *error = nil;
        NSData *encoded = [codec encrypt:_sessionState.contactsKey
                            withAuthData:_sessionState.authData
                               withError:&error];
        if (error == nil) {
            [realm beginWriteTransaction];
            config.whitelist = encoded;
            [realm commitWriteTransaction];
        }
        else {
            NSLog(@"Error while encoding whitelist: %@", error.localizedDescription);
        }
    }
    else {
        [realm beginWriteTransaction];
        config.whitelist = nil;
        [realm commitWriteTransaction];
    }

}

- (AccountConfig*)getConfig {

    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"accountName = %@", _sessionState.currentAccount];
    RLMResults<AccountConfig*> *configResults = [AccountConfig objectsWithPredicate:predicate];
    return [configResults firstObject];

}

- (NSInteger)getContactId:(NSString *)publicId {

    if (idMap.count == 0) {
        [self decodeIdMap:[self getConfig]];
    }
    NSNumber *cid = idMap[publicId];
    if (cid != nil) {
        return [cid integerValue];
    }
    else {
        return NSNotFound;
    }

}

- (NSString*)getContactPolicy {
    AccountConfig *config = [self getConfig];
    return config.contactPolicy;
}

- (NSString*)getNickname {
    AccountConfig *config = [self getConfig];
    return config.nickname;
}

- (void)loadWhitelist {

    AccountConfig *config = [self getConfig];
    if (privateWhitelist.count == 0) {
        [self decodeWhitelist:config];
    }

}

- (NSInteger)newContactId {

    AccountConfig *config = [self getConfig];
    [idLock lock];
    NSInteger contactId = config.contactId;
    
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm beginWriteTransaction];
    config.contactId = contactId + 1;
    [realm commitWriteTransaction];
    [idLock unlock];
    
    return contactId;
    
}

- (NSInteger)newMessageId {
    
    AccountConfig *config = [self getConfig];
    [idLock lock];
    NSInteger messageId = config.messageId;
    
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm beginWriteTransaction];
    config.messageId = messageId + 1;
    [realm commitWriteTransaction];
    [idLock unlock];

    return messageId;
    
}

- (void)setContactPolicy:(NSString *)policy {
    
    AccountConfig *config = [self getConfig];
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm beginWriteTransaction];
    config.contactPolicy = policy;
    [realm commitWriteTransaction];
    
}

- (void)setNickname:(NSString *)nickname {

    AccountConfig *config = [self getConfig];
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm beginWriteTransaction];
    config.nickname = nickname;
    [realm commitWriteTransaction];

}

- (void)startNewSession:(SessionState *)state {

    _sessionState = state;
    [idMap removeAllObjects];
    [privateWhitelist removeAllObjects];

}

@end

File: pippip-native-ios/Support/Configurator.swift
//
//  Configurator.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 5/22/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import RealmSwift
import CocoaLumberjack

struct Entity {
    
    var publicId: String
    var directoryId: String?
    
    init() {
        publicId = ""
    }
    
    init(publicId: String, directoryId: String?) {
        self.publicId = publicId
        self.directoryId = directoryId
    }

}

class Configurator: NSObject {

    static let currentVersion: Float = 2.3

    var whitelist: [Entity] {
        return privateWhitelist
    }
    private var privateWhitelist = [Entity]()
    var authenticated: Bool {
        get {
            let config = getConfig()
            return config.authenticated
        }
        set {
            let config = getConfig()
            let realm = try! Realm()
            try! realm.write {
                config.authenticated = newValue
                config.version = Configurator.currentVersion
            }
        }
    }
    var useLocalAuth: Bool {
        get {
            let config = getConfig()
            return config.useLocalAuth
        }
        set {
            let config = getConfig()
            let realm = try! Realm()
            try! realm.write {
                config.useLocalAuth = newValue
                config.version = Configurator.currentVersion
            }
        }
    }
    var showIgnoredContacts: Bool {
        get {
            let config = getConfig()
            return config.showIgnoredContacts
        }
        set {
            let config = getConfig()
            let realm = try! Realm()
            try! realm.write {
                config.showIgnoredContacts = newValue
                config.version = Configurator.currentVersion
            }
        }
    }
    var contactPolicy: String {
        get {
            let config = getConfig()
            return config.contactPolicy
        }
        set {
            let config = getConfig()
            let realm = try! Realm()
            try! realm.write {
                config.contactPolicy = newValue
                config.version = Configurator.currentVersion
            }
        }
    }
    var directoryId: String? {
        get {
            let config = getConfig()
            return config.directoryId
        }
        set {
            let config = getConfig()
            let realm = try! Realm()
            try! realm.write {
                config.directoryId = newValue
                config.version = Configurator.currentVersion
            }
        }
    }
    var statusUpdates: Int {
        get {
            let config = getConfig()
            return config.statusUpdates
        }
        set {
            let config = getConfig()
            let realm = try! Realm()
            try! realm.write {
                config.statusUpdates = newValue
                config.version = Configurator.currentVersion
            }
        }
    }
    var uuid: String {
        get {
            let config = getConfig()
            return config.uuid
        }
        set {
            let config = getConfig()
            let realm = try! Realm()
            try! realm.write {
                config.uuid = newValue
                config.version = Configurator.currentVersion
            }
        }
    }
    var autoAccept: Bool {
        get {
            let config = getConfig()
            return config.autoAccept
        }
        set {
            let config = getConfig()
            let realm = try! Realm()
            try! realm.write {
                config.autoAccept = newValue
                config.version = Configurator.currentVersion
            }
        }
    }
    var v2FirstRun: Bool {
        get {
            let config = getConfig()
            return config.v2FirstRun
        }
        set {
            let config = getConfig()
            let realm = try! Realm()
            try! realm.write {
                config.v2FirstRun = newValue
                config.version = Configurator.currentVersion
            }
        }
    }
    
    var sessionState = SessionState()

    func addWhitelistEntry(_ entity: Entity) throws {

        let config = getConfig()
        if privateWhitelist.isEmpty {
            try decodeWhitelist(config)
        }
        privateWhitelist.append(entity)
        try encodeWhitelist(config)
        
    }

    func decodeWhitelist(_ dbConfig: AccountConfig) throws {
        
        guard let dbWhitelist = dbConfig.whitelist else { return }
        let codec = CKGCMCodec(data: dbWhitelist)
        try codec.decrypt(sessionState.contactsKey!, withAuthData: sessionState.authData!)
        let count = codec.getLong()
        while whitelist.count < count {
            var entity = Entity()
            let directoryId = codec.getString()
            if directoryId.utf8.count > 0 {
                entity.directoryId = directoryId
            }
            entity.publicId = codec.getString()
            privateWhitelist.append(entity)
        }
        
    }
    
    func deleteWhitelistEntry(_ publicId: String) throws -> Int {

        let config = getConfig()
        if whitelist.isEmpty {
            try decodeWhitelist(config)
        }
        if let index = whitelist.index(where: { (entry) -> Bool in
            return publicId == entry.publicId
        }) {
            privateWhitelist.remove(at: index)
            try encodeWhitelist(config)
            return index
        }
        else {
            print("Whitelist entry for \(publicId) not found")
            return NSNotFound
        }

    }

    func encodeWhitelist(_ config: AccountConfig) throws {
        
        let realm = try Realm()
        if !whitelist.isEmpty {
            let codec = CKGCMCodec()
            codec.putLong(Int64(whitelist.count))
            for entity in whitelist {
                codec.put(entity.directoryId ?? "")
                codec.put(entity.publicId)
            }
            let encoded = codec.encrypt(sessionState.contactsKey!, withAuthData: sessionState.authData!)
            if encoded == nil {
                throw CryptoError(error: codec.lastError!)
            }
            try realm.write {
                config.whitelist = encoded
                config.version = Configurator.currentVersion
            }
        }
        else {
            try realm.write {
                config.whitelist = nil
                config.version = Configurator.currentVersion
            }
        }
        
    }

    private func getConfig() -> AccountConfig {
        
        do {
            let realm = try Realm()
            if let config = realm.objects(AccountConfig.self).first {
                return config
            }
            else {
                return AccountConfig()
            }
        }
        catch {
            DDLogError("Error retrieving configuration: \(error.localizedDescription)")
            return AccountConfig()
        }
        
    }
    
    func loadWhitelist() throws {

        privateWhitelist.removeAll()
        let config = getConfig()
        guard let _ = config.whitelist else { return }
        try decodeWhitelist(config)

    }

    func newContactId() ->Int {

        let config = getConfig()
        let contactId = config.currentContactId
        
        let realm = try! Realm()
        try! realm.write {
            config.currentContactId = contactId + 1
            config.version = Configurator.currentVersion
        }
        
        return contactId

    }

    func newMessageId() -> Int64 {

        let config = getConfig()
        let messageId = config.currentMessageId
        
        let realm = try! Realm()
        try! realm.write {
            config.currentMessageId = messageId + 1
            config.version = Configurator.currentVersion
        }
        
        return messageId

    }

    func whitelistIndexOf(publicId: String) -> Int? {

        do {
            if privateWhitelist.count == 0 {
                try loadWhitelist()
            }
            return privateWhitelist.index(where: {(entry) -> Bool in
                return publicId == entry.publicId
            })
        }
        catch {
            DDLogError("Error loading whitelist: \(error.localizedDescription)")
            return nil
        }

    }

}
